### PR TITLE
[SPARK-12186] [WEB UI] Send the complete request URI including the query string when redirecting.

### PR DESCRIPTION
**Pull request by rohitagarwal003**

### CODE REVIEW

**Comment by mindprince** [comment]
cc - @vanzin

**Comment by mindprince** [comment]
cc - @srowen, @andrewor14 Can you please ask Jenkins to run the tests. Thanks!

**Comment by andrewor14** [comment]
ok to test

**Comment by mindprince** [comment]
Can someone review this or if it seems okay, merge it?

File: core/src/main/scala/org/apache/spark/deploy/history/HistoryServer.scala [hunk 1]
```diff
@@ -103,7 +103,11 @@ class HistoryServer(
       // Note we don't use the UI retrieved from the cache; the cache loader above will register
       // the app's UI, and all we need to do is redirect the user to the same URI that was
       // requested, and the proper data should be served at that point.
-      res.sendRedirect(res.encodeRedirectURL(req.getRequestURI()))
+      if (req.getQueryString == null) {
```
Comment (srowen) [inline core/src/main/scala/org/apache/spark/deploy/history/HistoryServer.scala]: I think this is fine, but maybe slightly less repetitious like ...

```
val requestURI = req.getRequestURI + (if (req.getQueryString == null) "" else "?" + req.getQueryString)
res.sendRedirect(res.encodeRedirectURL(requestURI))

Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/deploy/history/HistoryServer.scala]: or even
```
val requestURI = req.getRequestURI + req.getQueryString.map("?" + _).getOrElse("")
```

File: core/src/main/scala/org/apache/spark/deploy/history/HistoryServer.scala [hunk 2]
```diff
@@ -103,7 +103,11 @@ class HistoryServer(
       // Note we don't use the UI retrieved from the cache; the cache loader above will register
       // the app's UI, and all we need to do is redirect the user to the same URI that was
       // requested, and the proper data should be served at that point.
```
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/deploy/history/HistoryServer.scala]: can you expand on this comment, summarizing SPARK-12186 in 1 sentence?

**Comment by andrewor14** [comment]
LGTM once you address the comments.

File: core/src/main/scala/org/apache/spark/deploy/history/HistoryServer.scala (hunk 1 above)
Comment (mindprince) [inline core/src/main/scala/org/apache/spark/deploy/history/HistoryServer.scala]: @andrewor14 Can't use `getOrElse`:
```
value getOrElse is not a member of scala.collection.immutable.IndexedSeq[String]
```

Comment (mindprince) [inline core/src/main/scala/org/apache/spark/deploy/history/HistoryServer.scala]: @srowen I think having an `if` statement like that makes it a little harder to read. How about the following, it has lesser repetition.
```
val requestURI = if (req.getQueryString == null) {
  req.getRequestURI
} else {
  req.getRequestURI + "?" + req.getQueryString
}
res.sendRedirect(res.encodeRedirectURL(requestURI))
```

Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/deploy/history/HistoryServer.scala]: what? Isn't `req.getQueryString` a string?

Comment (mindprince) [inline core/src/main/scala/org/apache/spark/deploy/history/HistoryServer.scala]: It is.

Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/deploy/history/HistoryServer.scala]: oh sorry, I meant
```
val requestURI = req.getRequestURI + Option(req.getQueryString).map("?" + _).getOrElse("")
```

File: core/src/main/scala/org/apache/spark/deploy/history/HistoryServer.scala (hunk 2 above)
Comment (mindprince) [inline core/src/main/scala/org/apache/spark/deploy/history/HistoryServer.scala]: Done.

File: core/src/main/scala/org/apache/spark/deploy/history/HistoryServer.scala (hunk 1 above)
Comment (mindprince) [inline core/src/main/scala/org/apache/spark/deploy/history/HistoryServer.scala]: Updated.

**Comment by andrewor14** [comment]
retest this please

**Comment by andrewor14** [comment]
Merged into master 1.6

**Pull request closed without merging**